README.md
[![banner](https://raw.githubusercontent.com/oceanprotocol/art/master/github/repo-banner%402x.png)](https://oceanprotocol.com)

# osmosis-azure-driver

> 💧 Osmosis Azure Driver Implementation
> [oceanprotocol.com](https://oceanprotocol.com)

[![Build Status](https://travis-ci.com/oceanprotocol/osmosis-azure-driver.svg)](https://travis-ci.com/oceanprotocol/osmosis-azure-driver)
[![PyPI](https://img.shields.io/pypi/v/osmosis-azure-driver.svg)](https://pypi.org/project/osmosis-azure-driver/)
[![GitHub contributors](https://img.shields.io/github/contributors/oceanprotocol/osmosis-azure-driver.svg)](https://github.com/oceanprotocol/osmosis-azure-driver/graphs/contributors)

---

## Table of Contents

- [Setup](#setup)
- [Code Style](#code-style)
- [Testing](#testing)
- [New Version](#new-version)
- [License](#license)

---

## Setup

To use Azure Storage with Brizo, you must set up some Azure Storage and then tell Brizo about your Azure Storage account by setting some Brizo configuration settings (either in a Brizo config file or in some environment variables). For details, see:

- [the README.md file in the Brizo repository](https://github.com/oceanprotocol/brizo/blob/develop/README.md) and
- [the tutorial about how to set up Azure Storage for use with Ocean Protocol](https://docs.oceanprotocol.com/tutorials/azure-for-brizo/)

## Code Style

Information about our Python code style is documented in the [python-developer-guide](https://github.com/oceanprotocol/dev-ocean/blob/master/doc/development/python-developer-guide.md)
and the [python-style-guide](https://github.com/oceanprotocol/dev-ocean/blob/master/doc/development/python-style-guide.md).

## Testing

Automatic tests are set up via Travis, executing `tox`.
Our tests use the pytest framework.

## New Version

The `bumpversion.sh` script helps to bump the project version. You can execute the script using as first argument {major|minor|patch} to bump accordingly the version.

## License

```text
Copyright 2018 Ocean Protocol Foundation Ltd.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```

Note: Installing this package (osmosis-azure-driver) installs the azure package which has an MIT license. Installing the azure package installs the certifi package which has an MPL-2.0 license. The certifi source code can be found at https://github.com/certifi/python-certifi . Installing the azure package also installs the chardet package which might have licensing that is incompatible with the MIT license (and the Apache-2.0 license). We have opened [an issue](https://github.com/Azure/azure-sdk-for-python/issues/4671) on the azure-sdk-for-python repository to let them know about the potential licensing conflict and to resolve it if necessary.


CHANGELOG.md
History
=======

0.0.4 (January 23rd, 2019)
---------------------------
* Fix error in config.
* Update README and license

0.0.2 (November 20th, 2018)
---------------------------
* Fix bug in generate sign url.

0.0.1 (November 13th, 2018)
---------------------------
* Plugin for manage data services
* Some computing capabilities.

tests/test_data_plugin.py
#  Copyright 2018 Ocean Protocol Foundation
#  SPDX-License-Identifier: Apache-2.0

from osmosis_azure_driver.data_plugin import Plugin
from osmosis_azure_driver.data_plugin import _parse_url
import os

osmo = Plugin('./tests/osmosis.ini')


def test_copy_file():
    assert osmo.type() == 'Azure'


# To run this test you need to login with your credentials through az login
def test_list():
    osmo.upload('./LICENSE', 'https://testocnfiles.blob.core.windows.net/ocn-hackaton/license_copy')
    osmo.download('https://testocnfiles.blob.core.windows.net/ocn-hackaton/license_copy', 'license_copy')
    assert open('license_copy').read() == open('./LICENSE').read()
    assert 'license_copy' in osmo.list('ocn-hackaton', True, 'testocnfiles')
    assert osmo.generate_url('https://testocnfiles.blob.core.windows.net/ocn-hackaton/license_copy')
    osmo.delete('https://testocnfiles.blob.core.windows.net/ocn-hackaton/license_copy')
    os.remove('license_copy')


def test_files_share():
    osmo.upload('./LICENSE', 'https://testocnfiles.file.core.windows.net/osmosis/license_copy')
    osmo.download('https://testocnfiles.file.core.windows.net/osmosis/license_copy', 'license_copy')
    assert open('license_copy').read() == open('./LICENSE').read()
    assert osmo.generate_url('https://testocnfiles.file.core.windows.net/osmosis/license_copy')
    assert 'license_copy' in osmo.list('osmosis', False, 'testocnfiles')
    osmo.delete('https://testocnfiles.file.core.windows.net/osmosis/license_copy')
    os.remove('license_copy')


def test_split_url():
    url = 'https://testocnfiles.blob.core.windows.net/mycontainer/myblob'
    parse_url = _parse_url(url)
    assert parse_url.account == 'testocnfiles'
    assert parse_url.container_or_share_name == 'mycontainer'
    assert parse_url.file == 'myblob'


def test_parse_file_url():
    url = 'https://testocnfiles.file.core.windows.net/compute/subfolder/data.txt'
    parse_url = _parse_url(url)
    assert parse_url.account == 'testocnfiles'
    assert parse_url.container_or_share_name == 'compute'
    assert parse_url.path == 'subfolder'
    assert parse_url.file == 'data.txt'


tests/test_computing_plugin.py
#  Copyright 2018 Ocean Protocol Foundation
#  SPDX-License-Identifier: Apache-2.0

from osmosis_azure_driver.data_plugin import Plugin as DataPlugin
from osmosis_azure_driver.computing_plugin import Plugin as ComputingPlugin
from osmosis_driver_interface.utils import parse_config


def test_compute_on_cloud():
    config = parse_config("./tests/osmosis.ini")
    osm_data = DataPlugin('./tests/osmosis.ini')
    osm_computing = ComputingPlugin('./tests/osmosis.ini')
    elements_before_compute = len(osm_data.list(config.get('azure.share.output'),
                                                False,
                                                config.get('azure.account.name')))
    result_file = osm_computing.exec_container(asset_url="data.txt",
                                               algorithm_url="algo.py",
                                               resource_group_name=config.get(
                                                   'azure.resource_group'),
                                               account_name=config.get('azure.account.name'),
                                               account_key=config.get('azure.account.key'),
                                               share_name_input=config.get(
                                                   'azure.share.input'),
                                               share_name_output=config.get(
                                                   'azure.share.output'),
                                               location=config.get('azure.location'),
                                               # input_mount_point=data.get('input_mount_point'),
                                               # output_mount_point=data.get('output_mount_point'),
                                               docker_image='python:3.6-alpine',
                                               memory=1.5,
                                               cpu=1)
    assert len(osm_data.list(config.get('azure.share.output'),
                             False,
                             config.get('azure.account.name'))) == elements_before_compute + 1
    osm_data.delete('https://testocnfiles.file.core.windows.net/output/' + result_file)


osmosis_azure_driver/log.py
#  Copyright 2018 Ocean Protocol Foundation
#  SPDX-License-Identifier: Apache-2.0

import yaml
import os
import logging.config
import logging
import coloredlogs


def setup_logging(default_path='logging.yaml', default_level=logging.INFO, env_key='LOG_CFG'):
    """Logging Setup"""
    path = default_path
    value = os.getenv(env_key, None)
    if value:
        path = value
    if os.path.exists(path):
        with open(path, 'rt') as f:
            try:
                config = yaml.safe_load(f.read())
                logging.config.dictConfig(config)
                coloredlogs.install()
            except Exception as e:
                print(e)
                print('Error in Logging Configuration. Using default configs')
                logging.basicConfig(level=default_level)
                coloredlogs.install(level=default_level)
    else:
        logging.basicConfig(level=default_level)
        coloredlogs.install(level=default_level)
        print('Failed to load configuration file. Using default configs')


osmosis_azure_driver/data_plugin.py
#  Copyright 2018 Ocean Protocol Foundation
#  SPDX-License-Identifier: Apache-2.0

import os
import logging
from datetime import datetime, timedelta
from azure.storage.blob import BlobPermissions
from azure.storage.blob import BlockBlobService
from azure.storage.file import FileService
from azure.common.cloud import get_cli_active_cloud
from azure.mgmt.storage import StorageManagementClient
from azure.mgmt.resource import ResourceManagementClient
from azure.common.credentials import get_azure_cli_credentials, ServicePrincipalCredentials
from osmosis_driver_interface.exceptions import OsmosisError
from osmosis_driver_interface.data_plugin import AbstractPlugin
from osmosis_azure_driver.utils import _parse_url
from osmosis_azure_driver.config import Config

class Plugin(AbstractPlugin):

    def __init__(self, config=None):
        """Initialize a :class:`~.Plugin`.
        """
        self.logger = logging.getLogger('Plugin')
        logging.basicConfig(level=logging.INFO)
        try:

            subscription_id = os.environ.get('AZURE_SUBSCRIPTION_ID')
            credentials = self._login_azure_app_token()
            # Create a Resource Management client
            self.resource_client = ResourceManagementClient(credentials, subscription_id)
            self.storage_client = StorageManagementClient(credentials, subscription_id)
        except Exception:
            logging.error('Credentials were not valid or were not found.')
            raise OsmosisError
        # self.resource_group_name = config.get('osmosis', 'azure.resource_group')  # OceanProtocol
        self.config=Config(config)
        self.resource_group_name = self.config.resource_group_name  # OceanProtocol

    @staticmethod
    def _login_azure_app_token(client_id=None, client_secret=None, tenant_id=None):
        """
        Authenticate APP using token credentials:
        https://docs.microsoft.com/en-us/python/azure/python-sdk-azure-authenticate?view=azure-python
        :return: ~ServicePrincipalCredentials credentials
        """
        client_id = os.getenv('AZURE_CLIENT_ID') if not client_id else client_id
        client_secret = os.getenv('AZURE_CLIENT_SECRET') if not client_secret else client_secret
        tenant_id = os.getenv('AZURE_TENANT_ID') if not tenant_id else tenant_id
        credentials = ServicePrincipalCredentials(
            client_id=client_id,
            secret=client_secret,
            tenant=tenant_id,
        )
        return credentials

    def _login_azure_cli(self):
        """
        Authenticate APP using az cli interactive login:
        https://docs.microsoft.com/en-us/python/azure/python-sdk-azure-authenticate?view=azure-python
        :return: credentials
        """
        return self._get_azure_cli_credentials()['credentials']

    def type(self):
        """str: the type of this plugin (``'Azure'``)"""
        return "Azure"

    def upload(self, local_file, remote_file):
        """Upload file to the cloud. The azure url format is https://myaccount.blob.core.windows.net/mycontainer/myblob.
         Args:
             local_file(str): The path of the file to be copied.
             remote_file(str): The destination path where the file is going to be allocated.
         Raises:
             :exc:`~..OsmosisError`: if the file is not uploaded correctly.
        """
        return self.copy(local_file, remote_file)

    def download(self, remote_file, local_file):
        """Download file from the cloud. The azure url format is https://myaccount.blob.core.windows.net/mycontainer/myblob.
         Args:
             remote_file(str): The path of the file to be copied.
             local_file(str): The destination path where the file is going to be allocated.
         Raises:
             :exc:`~..OsmosisError`: if the file is not downloaded correctly.
        """
        return self.copy(remote_file, local_file)

    def list(self, container_or_share_name, container=None, account=None):
        """List the blobs/files inside a container/share_name.
         Args:
             container_or_share_name(str): Name of the container/share_name where we want to list the blobs/files.
             container(bool): flag to know it you are listing files or blobs.
             account(str): The name of the storage account.
        """
        key = self.storage_client.storage_accounts.list_keys(self.resource_group_name, account).keys[0].value
        if container:
            bs = BlockBlobService(account_name=account, account_key=key)
            container_list = []
            for i in bs.list_blobs(container_or_share_name).items:
                container_list.append(i.name)
            return container_list
        elif not container:
            fs = FileService(account_name=account, account_key=key)
            container_list = []
            for i in fs.list_directories_and_files(container_or_share_name).items:
                container_list.append(i.name)
            return container_list
        else:
            raise ValueError("You have to pass a value for container param")

    def generate_url(self, remote_file):
        """Sign a remote file to distribute. The azure url format is https://myaccount.blob.core.windows.net/mycontainer/myblob.
         Args:
             remote_file(str): The blob that we want to sign.
        """
        parse_url = _parse_url(remote_file)
        key = self.storage_client.storage_accounts.list_keys(self.resource_group_name, parse_url.account).keys[0].value
        if parse_url.file_type == 'blob':
            bs = BlockBlobService(account_name=parse_url.account, account_key=key)

            sas_token = bs.generate_blob_shared_access_signature(parse_url.container_or_share_name,
                                                                 parse_url.file,
                                                                 permission=BlobPermissions.READ,
                                                                 expiry=datetime.utcnow() + timedelta(hours=24),
                                                                 )
            source_blob_url = bs.make_blob_url(container_name=parse_url.container_or_share_name,
                                               blob_name=parse_url.file,
                                               sas_token=sas_token)
            return source_blob_url
        elif parse_url.file_type == 'file':
            fs = FileService(account_name=parse_url.account, account_key=key)
            sas_token = fs.generate_file_shared_access_signature(share_name=parse_url.container_or_share_name,
                                                                 directory_name=parse_url.path,
                                                                 file_name=parse_url.file,
                                                                 permission=BlobPermissions.READ,
                                                                 expiry=datetime.utcnow() + timedelta(hours=24),
                                                                 )
            source_file_url = fs.make_file_url(share_name=parse_url.container_or_share_name,
                                               directory_name=parse_url.path,
                                               file_name=parse_url.file,
                                               sas_token=sas_token)
            return source_file_url
        else:
            raise ValueError("This azure storage type is not valid. It should be blob or file.")

    def delete(self, remote_file):
        """Delete file from the cloud. The azure url format is https://myaccount.blob.core.windows.net/mycontainer/myblob.
         Args:
             remote_file(str): The path of the file to be deleted.
         Raises:
             :exc:`~..OsmosisError`: if the file is not uploaded correctly.
        """
        if 'core.windows.net' not in remote_file:
            self.logger.error("Source or destination must be a azure storage url (format "
                              "https://myaccount.blob.core.windows.net/mycontainer/myblob")
            raise OsmosisError
        parse_url = _parse_url(remote_file)
        key = self.storage_client.storage_accounts.list_keys(self.resource_group_name, parse_url.account).keys[0].value
        if parse_url.file_type == 'blob':
            bs = BlockBlobService(account_name=parse_url.account, account_key=key)
            return bs.delete_blob(parse_url.container_or_share_name, parse_url.file)
        elif parse_url.file_type == 'file':
            fs = FileService(account_name=parse_url.account, account_key=key)
            return fs.delete_file(parse_url.container_or_share_name, parse_url.path, parse_url.file)
        else:
            raise ValueError("This azure storage type is not valid. It should be blob or file.")

    def copy(self, source_path, dest_path, account=None, group_name=None):
        """Copy file from a path to another path. The azure url format is https://myaccount.blob.core.windows.net/mycontainer/myblob.
         Args:
             source_path(str): The path of the file to be copied.
             dest_path(str): The destination path where the file is going to be allocated.
         Raises:
             :exc:`~..OsmosisError`: if the file is not uploaded correctly.
        """
        if 'core.windows.net' not in source_path and 'core.windows.net' not in dest_path:
            self.logger.error("Source or destination must be a azure storage url (format "
                              "https://myaccount.blob.core.windows.net/mycontainer/myblob")
            raise OsmosisError

        # Check if source exists and can read
        if 'core.windows.net' in source_path:
            parse_url = _parse_url(source_path)
            key = self.storage_client.storage_accounts.list_keys(self.resource_group_name, parse_url.account).keys[
                0].value
            if parse_url.file_type == 'blob':
                bs = BlockBlobService(account_name=parse_url.account, account_key=key)
                return bs.get_blob_to_path(parse_url.container_or_share_name, parse_url.file, dest_path)
            elif parse_url.file_type == 'file':
                fs = FileService(account_name=parse_url.account, account_key=key)
                return fs.get_file_to_path(parse_url.container_or_share_name, parse_url.path, parse_url.file, dest_path)
            else:
                raise ValueError("This azure storage type is not valid. It should be blob or file.")
        else:
            parse_url = _parse_url(dest_path)
            key = self.storage_client.storage_accounts.list_keys(self.resource_group_name, parse_url.account).keys[
                0].value
            if parse_url.file_type == 'blob':
                bs = BlockBlobService(account_name=parse_url.account, account_key=key)
                return bs.create_blob_from_path(parse_url.container_or_share_name, parse_url.file, source_path)
            elif parse_url.file_type == 'file':
                fs = FileService(account_name=parse_url.account, account_key=key)
                return fs.create_file_from_path(parse_url.container_or_share_name, parse_url.path, parse_url.file,
                                                source_path)
            else:
                raise ValueError("This azure storage type is not valid. It should be blob or file.")

    def create_directory(self, remote_folder, container=None):
        if container:
            return self.create_container(remote_folder)
        elif not container:
            return self.create_share_name()
        else:
            raise ValueError("You have to pass a value for container param")

    def create_container(self, remote_folder):
        parse_url = _parse_url(remote_folder)
        key = self.storage_client.storage_accounts.list_keys(self.resource_group_name, parse_url.account).keys[0].value
        bs = BlockBlobService(account_name=parse_url.account, account_key=key)
        return bs.create_container(container_name=remote_folder)

    def create_share_name(self, remote_folder):
        parse_url = _parse_url(remote_folder)
        key = self.storage_client.storage_accounts.list_keys(self.resource_group_name, parse_url.account).keys[0].value
        fs = FileService(account_name=parse_url.account, account_key=key)
        return fs.create_directory(share_name=parse_url.container_or_share_name, directory_name=parse_url.path)

    def retrieve_availability_proof(self):
        pass

    @staticmethod
    def _get_azure_cli_credentials():
        credentials, subscription_id = get_azure_cli_credentials()
        cloud_environment = get_cli_active_cloud()

        cli_credentials = {
            'credentials': credentials,
            'subscription_id': subscription_id,
            'cloud_environment': cloud_environment
        }
        return cli_credentials


osmosis_azure_driver/computing_plugin.py
#  Copyright 2018 Ocean Protocol Foundation
#  SPDX-License-Identifier: Apache-2.0

import logging
import os
import time

from azure.common.cloud import get_cli_active_cloud
from azure.common.credentials import get_azure_cli_credentials, ServicePrincipalCredentials
from azure.mgmt.containerinstance import ContainerInstanceManagementClient
from azure.mgmt.containerinstance.models import (ContainerGroup, Container, ResourceRequirements,
                                                 ResourceRequests,
                                                 OperatingSystemTypes, Volume, VolumeMount,
                                                 AzureFileVolume,
                                                 ContainerGroupRestartPolicy)
from azure.mgmt.resource import ResourceManagementClient
from osmosis_driver_interface.computing_plugin import AbstractPlugin
from osmosis_driver_interface.exceptions import OsmosisError


class Plugin(AbstractPlugin):

    def __init__(self, config=None, resource_group_name=None):
        """Initialize a :class:`~.Plugin`.
        """
        self.logger = logging.getLogger('Plugin')
        logging.basicConfig(level=logging.INFO)
        try:

            subscription_id = os.environ.get('AZURE_SUBSCRIPTION_ID')
            credentials = self._login_azure_app_token()
            # Create a Resource Management client
            self.resource_client = ResourceManagementClient(credentials, subscription_id)
            self.client = ContainerInstanceManagementClient(credentials, subscription_id)
        except Exception:
            logging.error('Credentials were not valid or were not found.')
            raise OsmosisError
        # self.resource_group_name = config.get('osmosis', 'azure.resource_group')  # OceanProtocol
        self.resource_group_name = resource_group_name  # OceanProtocol

    @staticmethod
    def _login_azure_app_token(client_id=None, client_secret=None, tenant_id=None):
        """
        Authenticate APP using token credentials:
        https://docs.microsoft.com/en-us/python/azure/python-sdk-azure-authenticate?view=azure-python
        :return: ~ServicePrincipalCredentials credentials
        """
        client_id = os.getenv('AZURE_CLIENT_ID') if not client_id else client_id
        client_secret = os.getenv('AZURE_CLIENT_SECRET') if not client_secret else client_secret
        tenant_id = os.getenv('AZURE_TENANT_ID') if not tenant_id else tenant_id
        credentials = ServicePrincipalCredentials(
            client_id=client_id,
            secret=client_secret,
            tenant=tenant_id,
        )
        return credentials

    def _login_azure_cli(self):
        """
        Authenticate APP using az cli interactive login:
        https://docs.microsoft.com/en-us/python/azure/python-sdk-azure-authenticate?view=azure-python
        :return: credentials
        """
        return self._get_azure_cli_credentials()['credentials']

    def type(self):
        """str: the type of this plugin (``'Azure'``)"""
        return "Azure"

    def create_vm(self, container_group=None):
        if container_group is not None:
            return container_group
        else:
            raise ValueError("You should provide a container group.")

    def _create_container_group(self, resource_group_name,
                                name,
                                location,
                                image,
                                memory,
                                cpu,
                                algorithm,
                                asset,
                                input_mount_point,
                                output_moint_point,
                                account_name,
                                account_key,
                                share_name_input,
                                share_name_output):
        # setup default values
        result_file = 'result-' + str(int(time.time()))
        command = ['python', input_mount_point + '/' + algorithm, input_mount_point + '/' + asset,
                   output_moint_point + '/' + result_file]
        environment_variables = None
        az_file_input = AzureFileVolume(share_name=share_name_input,
                                        storage_account_name=account_name,
                                        storage_account_key=account_key,
                                        )

        az_file_output = AzureFileVolume(share_name=share_name_output,
                                         storage_account_name=account_name,
                                         storage_account_key=account_key,
                                         )

        volume = [Volume(name=share_name_input, azure_file=az_file_input),
                  Volume(name=share_name_output, azure_file=az_file_output)]
        volume_mount = [VolumeMount(name=share_name_input, mount_path=input_mount_point),
                        VolumeMount(name=share_name_output, mount_path=output_moint_point)]

        # set memory and cpu
        container_resource_requests = ResourceRequests(memory_in_gb=memory, cpu=cpu)
        container_resource_requirements = ResourceRequirements(requests=container_resource_requests)

        container = Container(name=name,
                              image=image,
                              resources=container_resource_requirements,
                              command=command,
                              environment_variables=environment_variables,
                              volume_mounts=volume_mount,
                              )

        # defaults for container group
        cgroup_os_type = OperatingSystemTypes.linux

        cgroup = ContainerGroup(location=location,
                                containers=[container],
                                os_type=cgroup_os_type,
                                restart_policy=ContainerGroupRestartPolicy.never,
                                volumes=volume,
                                )

        self.client.container_groups.create_or_update(resource_group_name, name, cgroup)
        return result_file

    def exec_container(self,
                       asset_url,
                       algorithm_url,
                       resource_group_name,
                       account_name,
                       account_key,
                       location,
                       share_name_input='compute',
                       share_name_output='output',
                       docker_image='python:3.6-alpine',
                       memory=1.5,
                       cpu=1):
        """Prepare a docker image that will run in the cloud, mounting the asset and executing the algorithm.
        :param asset_url
        :param algorithm_url
        :param resource_group_name:
        :param account_name:
        :param account_key:
        :param share_name_input:
        :param share_name_output:
        :param location:
        """
        try:
            container_group_name = 'compute' + str(int(time.time()))
            result_file = self._create_container_group(resource_group_name=resource_group_name,
                                                       name=container_group_name,
                                                       image=docker_image,
                                                       location=location,
                                                       memory=memory,
                                                       cpu=cpu,
                                                       algorithm=algorithm_url,
                                                       asset=asset_url,
                                                       input_mount_point='/input',
                                                       output_moint_point='/output',
                                                       account_name=account_name,
                                                       account_key=account_key,
                                                       share_name_input=share_name_input,
                                                       share_name_output=share_name_output
                                                       )
            while self.client.container_groups.get(resource_group_name,
                                                   container_group_name).provisioning_state != 'Succeeded':
                logging.info("Waiting to resources ")
            while self.client.container_groups.get(resource_group_name, container_group_name). \
                    containers[0].instance_view.current_state.state != 'Terminated':
                logging.info("Waiting to terminate")
            self.delete_vm(container_group_name, resource_group_name)
            return result_file
        except Exception:
            logging.error("There was a problem executing your container")
            raise Exception

    def start_vm(self, instance_name):
        pass

    def stop_vm(self, instance_name):
        pass

    def run_command(self, instance_name, command):
        pass

    def delete_vm(self, container_group_name, resource_group_name=None):
        return self.client.container_groups.delete(resource_group_name, container_group_name)

    def status_vm(self, instance_name):
        pass

    def copy(self, instance_name, source_path, dest_path):
        pass

    def retrieve_computation_proof(self):
        pass

    def retrieve_vm_logs(self):
        pass

    def list_container_groups(self, resource_group_name):
        """Lists the container groups in the specified resource group.

        Arguments:
           aci_client {azure.mgmt.containerinstance.ContainerInstanceManagementClient}
                       -- An authenticated container instance management client.
           resource_group {azure.mgmt.resource.resources.models.ResourceGroup}
                       -- The resource group containing the container group(s).
        """
        print("Listing container groups in resource group '{0}'...".format(resource_group_name))

        container_groups = self.client.container_groups.list_by_resource_group(resource_group_name)

        for container_group in container_groups:
            print("  {0}".format(container_group.name))

    @staticmethod
    def _get_azure_cli_credentials():
        credentials, subscription_id = get_azure_cli_credentials()
        cloud_environment = get_cli_active_cloud()

        cli_credentials = {
            'credentials': credentials,
            'subscription_id': subscription_id,
            'cloud_environment': cloud_environment
        }
        return cli_credentials
